### PR TITLE
Refactor web and mobile docs

### PR DESCRIPTION
**Pull request by natke**
This PR contains the changes outlined in the review here #9398.

Staged here: https://natke.github.io/onnxruntime/docs/

**Summary of changes**

Web tutorials (currently just one): 
- https://natke.github.io/onnxruntime/docs/tutorials/web/classify-images-nextjs-github-template.html

Mobile tutorials:
- https://natke.github.io/onnxruntime/docs/tutorials/mobile/deploy-ios.html
- https://natke.github.io/onnxruntime/docs/tutorials/mobile/deploy-android.html

Pre-built package info moved into install doc:
- https://natke.github.io/onnxruntime/docs/install/#install-on-web-and-mobile

Moved overview section for mobile into Get Started with Mobile:
- https://natke.github.io/onnxruntime/docs/get-started/with-mobile.html

Moved custom build info into Build section (with links from tutorials to here):
- https://natke.github.io/onnxruntime/docs/build/custom.html

Moved reduced operator config file specification into Reference, with links from build and tutorials:
- https://natke.github.io/onnxruntime/docs/reference/reduced-operator-config-file.html

Moved ONNX to ORT format conversion process into Reference, with links from build and tutorials:
- https://natke.github.io/onnxruntime/docs/reference/ort-format-model-conversion.html

Created Build Web Apps with ONNX Runtime:
(This section contains general instructions and considerations for building a web app with ORT. Ideally I would like to see us build out more scenarios that would encompass the different options presented in this doc).
- https://natke.github.io/onnxruntime/docs/reference/build-web-app.html

### CODE REVIEW

**Review by import-brain — APPROVED** [review]

File: docs/build/custom.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+A minimal build has the following limitations:
+
+* No support for ONNX format models, that is model must be converted to ORT format
+* No support for runtime optimizations. Optimizations are performed during conversion to ORT format
+* Limited support for runtime partitioning (assigning nodes in a model to specific execution providers). Execution providers that statically register kernels (e.g. ONNX Runtime CPU Execution Provider) are supported by default. All execution providers that will be used at runtime MUST be registered when creating the ORT format model
+    - Execution providers that compile nodes are optionally supported
+      - currently this is limited to the NNAPI and CoreML Execution Providers
+
+We do not currently offer backwards compatibility guarantees for ORT format models, as we will be expanding the capabilities in the short term and may need to update the internal format in an incompatible manner to accommodate these changes. You may need to regenerate the ORT format models to use with a future version of ONNX Runtime. Once the feature set stabilizes we will provide backwards compatibility guarantees.
```
Comment (skottmckay) [inline docs/build/custom.md:66]: >We do not currently offer backwards compatibility guarantees for ORT format models, as we will be expanding the capabilities in the short term and may need to update the internal format in an incompatible manner to accommodate these changes. [](http://example.com/codeflow?start=0&length=241)

Can we update this to say that we now offer a backwards compatibility guarantee?

File: docs/build/custom.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+To build a custom ONNX Runtime package, the [build from source](./index.md) instructions apply, with some extra options that are specified below.
+
+## Contents
+{: .no_toc }
+
+* TOC placeholder
+{:toc}
+
+## Reduce operator set
+
```
Comment (skottmckay) [inline docs/build/custom.md:27]: The 'set' might be confusing here given onnx has the notion of an 'operator set' (generally called the opset). 

'Reduce operator kernels' might be clearer as a kernel is the implementation of the operator for the opset the model requires.

File: docs/reference/ort-format-model-conversion.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+import * as ort from "onnxruntime-web";
+
+const session = await ort.InferenceSession.create("<path to model>");
+```
+
+#### Load ORT format model from an in-memory byte array
+
+If a session is created using an input byte array containing the ORT format model data, by default we will copy the model bytes at the time of session creation to ensure the model bytes buffer is valid.
+
+You may also enable the option to use the model bytes directly by setting the Session Options config `session.use_ort_model_bytes_directly` to `1`, this may reduce the peak memory usage of ONNX Runtime Mobile, you will need to guarantee that the model bytes are valid throughout the lifespan of the ORT session using the model bytes. For ONNX Runtime Web, this option is set by default.
```
Comment (skottmckay) [inline docs/reference/ort-format-model-conversion.md:220]: >You may also enable the option to use the model bytes directly by setting the Session Options config `session.use_ort_model_bytes_directly` to `1`, this may reduce the peak memory usage of ONNX Runtime Mobile, you will need to guarantee that the model bytes are valid throughout the lifespan of the ORT session using the model bytes. [](http://example.com/codeflow?start=0&length=333)

nit: maybe split this sentence up

File: docs/reference/operators/mobile_package_op_type_support_1.9.md [hunk 4]
```diff
@@ -1,8 +1,8 @@
 ---
-title: ORT 1.9 Mobile Package Operator and Type Support
-parent: Pre-Built Package
-grand_parent: ONNX Runtime Mobile
-nav_exclude: true
+title: ORT 1.9 Mobile Package Operators
+parent: Operators
```
Comment (skottmckay) [inline docs/reference/operators/mobile_package_op_type_support_1.9.md:3]: Should we add 1.10 while we're at it? The release branch is cut so the operators set doco can be generated on it.

File: docs/build/custom.md (hunk 1 above)
Comment (natke) [reply docs/build/custom.md:66]: Yes, if we do, that's great.

File: docs/reference/ort-format-model-conversion.md (hunk 3 above)
Comment (natke) [reply docs/reference/ort-format-model-conversion.md:220]: Agreed - there's some serious comma splicing in there. Well spotted!

File: docs/reference/operators/mobile_package_op_type_support_1.9.md (hunk 4 above)
Comment (natke) [reply docs/reference/operators/mobile_package_op_type_support_1.9.md:3]: Yes, we should! Is this auto-generated? Where would I find it?

Comment (skottmckay) [reply docs/reference/operators/mobile_package_op_type_support_1.9.md:3]: In the ORT enlistment /tools/python/gen_ort_mobile_pkg_doc.py will update /docs/ORTMobilePackageOperatorTypeSupport.md

Comment (skottmckay) [reply docs/reference/operators/mobile_package_op_type_support_1.9.md:3]: Updated version is also checked in to master so you can copy from /docs/ORTMobilePackageOperatorTypeSupport.md

**Comment by skottmckay** [comment]
Can we also update the main page to add C# as an API for Android and iOS (via Xamarin)? Could be a separate PR (not clear to me how to do that though - the index.html is a bit cryptic).

File: docs/build/custom.md (hunk 1 above)
Comment (natke) [reply docs/build/custom.md:66]: Fixed

File: docs/build/custom.md (hunk 2 above)
Comment (natke) [reply docs/build/custom.md:27]: Fixed

File: docs/reference/ort-format-model-conversion.md (hunk 3 above)
Comment (natke) [reply docs/reference/ort-format-model-conversion.md:229]: Fixed

File: docs/reference/operators/mobile_package_op_type_support_1.9.md (hunk 4 above)
Comment (natke) [reply docs/reference/operators/mobile_package_op_type_support_1.9.md:3]: Added

**Comment by natke** [comment]
> Can we also update the main page to add C# as an API for Android and iOS (via Xamarin)? Could be a separate PR (not clear to me how to do that though - the index.html is a bit cryptic).

We will add this as a separate PR

File: docs/tutorials/web/index.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+To see an example of this flow in practice, you can follow the steps in the following tutorial to [build a web application to classify images using Next.js](classify-images-nextjs-github-template.md).
+
+For more detail on the steps below, see the [build a web application](../../reference/build-web-app.md) with ONNX Runtime reference guide.
+
+## ONNX Runtime web application development flow
+
+1. Choose deployment target and ONNX Runtime package
+
+   ONNX Runtime Web can be integrated into your web application in a number of different ways depending on the requirements of your application.
```
Comment (fs-eire) [inline docs/tutorials/web/index.md:20]: not accurate. "ONNX Runtime Web"=>"ONNX Runtime libraries". Only the first scenario (Inference in browser) uses "ONNX Runtime Web". The other scenarios uses other libraries of ORT.

Comment (fs-eire) [reply docs/tutorials/web/index.md:20]: There is another important scenario missing in this list: Electron.js

Electron is an important scenario, and we should mention it. (VS Code, for example, is built up using Electron)

ONNX Runtime Web may be useful in Electron based app, depending on the requirement

File: docs/tutorials/web/index.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+   * Inference on server in JavaScript. Use the `onnxruntime-node` package.
+
+     Your application may have constraints that means it is better to perform inference server side.
+
+     * **The model is too large** and requires higher hardware specs. In order to do inference on the client you need to have a model that is small enough to run efficiently on less powerful hardware.
+     * You don't want the model to be downloaded onto the device.
+
+     You can also use the onnxruntime-node package in the backend of an elctron app.
+
+   * Inference on server using other language APIs. Use the `onnxruntime` packages for C/C++ and other languages.
```
Comment (fs-eire) [inline docs/tutorials/web/index.md:44]: this is out of scope. but don't highlight "onnxruntime" as if it is a name of a NPM package here. This may confuse users. 

users don't need any NPM package in this scenario (they may try to look up to "onnxruntime" package in NPM repository, and what makes it worse is that there is such a package named "onnxruntime" -- the deprecated package for onnxruntime nodejs binding)

File: docs/tutorials/web/index.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+     * **If the application is performance critical**. If the onnxruntime-node server-side JavaScript library does not deliver the performance that your application needs then you can use the native ONNX Runtime libraries.
```
Comment (fs-eire) [inline docs/tutorials/web/index.md:46]: This is incorrect. Node.js binding is of very good performance, and it is the best choice for Node.js server solution.

The only reason why they don't use Node.js binding in their server is they do not use Node.js.

File: docs/tutorials/web/index.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+   * [Use Azure Machine Learning Designer to create a custom model](https://docs.microsoft.com/en-us/azure/machine-learning/concept-designer)
+   * [Build your own model with PyTorch.](https://docs.microsoft.com/learn/paths/pytorch-fundamentals/)
+
+3. How do I bootstrap my app development?
+
+   Bootstrap your web application according in your web framework of choice e.g. vuejs, reactjs, angularjs.
+
+   a. Add the ONNX Runtime dependency
+   b. Consume the onnxruntime-web API in your application
+   c. Add pre and post processing appropriate to your application and model
```
Comment (fs-eire) [inline docs/tutorials/web/index.md:69]: now they are showing in single line. please fix it

Comment (fs-eire) [reply docs/tutorials/web/index.md:69]: also, can add link to build-web-app

for example, build-web-app.md#add-onnx-runtime-web-as-dependency

File: docs/tutorials/web/index.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+4. How do I optimize my application?
+
+   The libraries and models mentioned in the previous steps can be optimized to meet memory and processing demands.
+
```
Comment (fs-eire) [inline docs/tutorials/web/index.md:74]: we can add one here:
 - tune session options ( including trying different EPs)

File: docs/reference/build-web-app.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+   Browser sends user's input to server, server inferences and gets the result and sends back to client.
+
+   Use native ONNX Runtime to get best performance.
+
+   To use Node.js as the server application, use onnxruntime-node (ONNX Runtime node.js binding) on the server.
+
+3. Electron
+
+   Electron uses a frontend (based on chromium, technically a browser core) and a backend (based on Node.js). Examples are VSCode, Teams before 2021.
```
Comment (fs-eire) [inline docs/reference/build-web-app.md:36]: remove the examples (or at least remove Teams)

File: docs/reference/build-web-app.md [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  
+## Pre and post processing
+
+Raw input is usually a string (for NLP model) or an image (for image model). They come from multiple forms and formats.
+
+### String inputs
+
+1. Use a tokenizer in JS/wasm to pre-process it to number data, create tensors from the data and feed to ORT for model inferencing.
+
+2. [example: OWA inline file suggestions]: use one or more custom ops to deal with strings. Build with the custom ops. The model can directly process string tensor inputs. Refer to the[onnxruntime-extensions](https://github.com/microsoft/onnxruntime-extensions) library, which contain a set of possible custom operators.
```
Comment (fs-eire) [inline docs/reference/build-web-app.md:103]: Not sure if this is appropriate for a public example to show

File: docs/tutorials/web/index.md (hunk 5 above)
Comment (natke) [reply docs/tutorials/web/index.md:20]: Electron is mentioned further down in the paragraph.

Would the sentence be accurate if it was worded thus:

```
ONNX Runtime can be integrated into your web application in a number of different ways depending on the requirements of your application.
```

My goal is to keep this customer scenario focused.

File: docs/tutorials/web/index.md (hunk 6 above)
Comment (natke) [reply docs/tutorials/web/index.md:44]: Fixed

File: docs/tutorials/web/index.md (hunk 7 above)
Comment (natke) [reply docs/tutorials/web/index.md:46]: Fixed

File: docs/tutorials/web/index.md (hunk 8 above)
Comment (natke) [reply docs/tutorials/web/index.md:69]: Fixed

File: docs/tutorials/web/index.md (hunk 9 above)
Comment (natke) [reply docs/tutorials/web/index.md:74]: Added

File: docs/reference/build-web-app.md (hunk 10 above)
Comment (natke) [reply docs/reference/build-web-app.md:36]: Removed

File: docs/tutorials/web/index.md (hunk 5 above)
Comment (natke) [reply docs/tutorials/web/index.md:20]: I'm thinking as I make these edits that it might be better to have a JavaScript developer page rather than this more generic one. But I think we should publish this and then move on to the improvements.

File: docs/tutorials/web/index.md (hunk 8 above)
Comment (fs-eire) [reply docs/tutorials/web/index.md]: we can add links to (b) and (c) as they are mapping to the sections in build-web-app.md

they are "#consume-onnxruntime-web-in-your-code" and "#pre-and-post-processing"

File: docs/reference/build-web-app.md (hunk 11 above)
Comment (natke) [reply docs/reference/build-web-app.md]: Removed example

File: docs/tutorials/web/index.md [hunk 12]
```diff
@@ -9,15 +9,17 @@ nav_order: 6
 
 ONNX Runtime Web enables you to run and deploy machine learning models in your web application using JavaScript APIs and libraries. This page outlines the general flow through the development process.
 
-To see an example of this flow in practice, you can follow the steps in the following tutorial to [build a web application to classify images using Next.js](classify-images-nextjs-github-template.md).
+You can also integrate machine learning into the server side of your web application with ONNX Runtime using other language language libraries, depending on your application development environment.
```
Comment (hanbitmyths) [inline docs/tutorials/web/index.md:12]: duplicate here: 'language language'

File: docs/tutorials/web/index.md [hunk 13]
```diff
@@ -64,9 +66,11 @@ For more detail on the steps below, see the [build a web application](../../refe
 
    Bootstrap your web application according in your web framework of choice e.g. vuejs, reactjs, angularjs.
 
-   a. Add the ONNX Runtime dependency
-   b. Consume the onnxruntime-web API in your application
-   c. Add pre and post processing appropriate to your application and model
+   1. [Add the ONNX Runtime dependency](../../reference/build-web-app.md#add-onnx-runtime-web-as-dependency)
+
+   1. Consume the onnxruntime-web API in your application
```
Comment (hanbitmyths) [inline docs/tutorials/web/index.md:71]: Link to '../../reference/build-web-app.html#consume-onnxruntime-web-in-your-code'

File: docs/tutorials/web/index.md [hunk 14]
```diff
@@ -64,9 +66,11 @@ For more detail on the steps below, see the [build a web application](../../refe
...
+
+   1. Add pre and post processing appropriate to your application and model
```
Comment (hanbitmyths) [inline docs/tutorials/web/index.md:73]: Link to '../../reference/build-web-app.html#pre-and-post-processing'

Comment (natke) [reply docs/tutorials/web/index.md]: Fixed

File: docs/tutorials/web/index.md (hunk 13 above)
Comment (natke) [reply docs/tutorials/web/index.md]: Added

File: docs/tutorials/web/index.md (hunk 12 above)
Comment (natke) [reply docs/tutorials/web/index.md]: Fixed. Well spotted!

**Review by hanbitmyths — APPROVED** [review]
LGTM
